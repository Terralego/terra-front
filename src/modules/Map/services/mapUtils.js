import bbox from '@turf/bbox';
import moize from 'moize';

import { PREFIXES } from './cluster';

export const PREV_STATE = {};
export const LAYER_TYPES_ORDER = ['background', 'raster', 'hillshade', 'heatmap', 'fill', 'fill-extrusion', 'line', 'circle', 'symbol'];

export const getRelatedLayers = moize({
  serializer: (map, layerId) => `${layerId}${map.getStyle().layers.map(id => id).join('')}`,
})((map, layerId) => {
  const regexp = new RegExp(`^${layerId}(-(${PREFIXES.join('|')}))?(-[0-9]+)?$`);
  return map.getStyle().layers
    .filter(({ id }) => id.match(regexp));
});

export function toggleLayerVisibility (map, layerId, visibility) {
  getRelatedLayers(map, layerId)
    .forEach(({ id }) => map.setLayoutProperty(id, 'visibility', visibility));
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  getRelatedLayers(map, layerId)
    .forEach(layer => {
      const property = getOpacityProperty(layer.type);
      if (property) {
        map.setPaintProperty(layer.id, property, opacity);
      }
    });
}

export const checkContraints = ({
  map,
  constraints = [],
  feature: { properties: { cluster } = {} } = {},
}) => {
  if (!constraints.length) return true;
  const currentZoom = map.getZoom();
  return constraints.reduce((prev, {
    minZoom = 0,
    maxZoom = Infinity,
    withLayers = [],
    isCluster,
  }) => {
    const checkZoom = currentZoom <= maxZoom && currentZoom >= minZoom;

    const checkLayers = withLayers.reduce((prevCheck, layer) => {
      const match = layer.match(/^(!)?(.+)$/);

      const visible = match[1] !== '!';
      const layerId = match[2];
      return prevCheck && visible
        ? getRelatedLayers(map, layerId).reduce((subCheck, { id }) =>
          (subCheck ||
          map.getLayoutProperty(id, 'visibility') === 'visible'), false)
        : getRelatedLayers(map, layerId).reduce((subCheck, { id }) =>
          (subCheck &&
          map.getLayoutProperty(id, 'visibility') === 'none'), true);
    }, true);

    const checkCluster = isCluster === undefined
      || !!cluster === !!isCluster;

    return prev || (checkZoom && checkLayers && checkCluster);
  }, false);
};

export function getInteractionsOnEvent ({
  eventType,
  map,
  point,
  interactions: eventInteractions,
  features = map.queryRenderedFeatures(point),
}) {
  let interactions = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteractions = eventInteractions.filter(({ id, trigger = 'click', constraints }) => {
      if (eventType !== (trigger === 'mouseover' ? 'mousemove' : trigger)) return false;

      const found = getRelatedLayers(map, id)
        .find(({ id: compatibleId }) => layerId === compatibleId);

      if (!found) {
        return false;
      }

      if (!constraints) {
        return true;
      }

      return checkContraints({ map, constraints, feature });
    });

    if (!foundInteractions.length) return false;

    interactions = {
      interactions: foundInteractions,
      feature,
      layerId,
    };

    return true;
  });

  return interactions;
}

const listenerWaiter = {};
export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  /**
   * mouseover is a terralego specific event which equals to a mousemove mapbox
   * event which watch when entering then leaving a feature
   */
  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  /**
   * Mouseleave events for mouseover triggers
   * /!\ this listener MUST be before the mousemove
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;
    const eventType = 'mouseleave';

    getRelatedLayers(map, id).forEach(({ id: realLayer }) => {
      map.on(eventType, realLayer, event => {
        const { features = [] } = PREV_STATE;
        const feature = features.find(({ layer: { id: layerId } }) => id === layerId);
        callback({ event, map, layerId: id, interaction, feature, eventType });
      });
    });
  });

  const listener = (e, eventType) => {
    const { target, point } = e;
    let features;
    if (eventType === 'mousemove') {
      features = map.queryRenderedFeatures(point);
      PREV_STATE.features = features;
    }
    const interactionsSpec = getInteractionsOnEvent({
      eventType,
      map: target,
      point,
      interactions,
      features,
    });

    if (!interactionsSpec) return;

    const { interactions: filteredInteractionsSpec, feature, layerId } = interactionsSpec;
    filteredInteractionsSpec.forEach(interaction =>
      callback({ event: e, map, layerId, feature, interaction, eventType }));
  };
  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      clearTimeout(listenerWaiter[eventType]);
      listenerWaiter[eventType] = setTimeout(() => listener(e, eventType), eventType === 'mousemove' ? 200 : 100);
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionsSpec = getInteractionsOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionsSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export function fitZoom ({ feature, map, padding = 0 }) {
  const features = feature.length ? feature : [feature];
  map.fitBounds(bbox({ type: 'FeatureCollection', features }), { padding });
}

export const getLayerBeforeId = (type, layers) => {
  const sameTypes = layers.filter(layer => type === layer.type);

  if (!sameTypes.length) return undefined;

  const lastOfSameType = sameTypes[sameTypes.length - 1];

  const pos = layers.findIndex(({ id }) => id === lastOfSameType.id) + 1;

  return layers[pos] && layers[pos].id;
};

export const getOrderedLayers = (layers, layersTypes = LAYER_TYPES_ORDER) => {
  const orderedLayers = [...layers];
  return orderedLayers.sort(({ type: typeA }, { type: typeB }) => {
    if (typeA === typeB) return 0;
    return layersTypes.indexOf(typeA) - layersTypes.indexOf(typeB);
  });
};

export default {
  LAYER_TYPES_ORDER,
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionsOnEvent,
  setInteractions,
  checkContraints,
  fitZoom,
  getLayerBeforeId,
  getOrderedLayers,
};
