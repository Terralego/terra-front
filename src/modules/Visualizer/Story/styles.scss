@import '../../../utils/variables.scss'; 

.storytelling {
  display: flex;
  flex-direction: column;
  min-height: calc(100vh - 140px);
  padding: 0 .75rem;
  padding-bottom: 1rem;
  &__content {
    flex: 1;
  }
  &__buttons {
    display: flex;
    font-size: 2rem;
    height: 1.3em;
    justify-content: center;
    &--justified {
      justify-content: space-between;
    }
  }
}

@media(min-width: $breakpoints-tablet) and (max-width: $breakpoints-desktop) and (orientation: landscape) {
  .storytelling {
    font-size: 1rem/$scaleRatio;
    line-height: 145%;

    h2 {
      line-height: 145%;
      font-size: 1.3rem/$scaleRatio;
    }

    &__buttons {
      font-size: 2rem/$scaleRatio;
    }
  }
}
