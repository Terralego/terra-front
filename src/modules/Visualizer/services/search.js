import elasticsearch from 'elasticsearch';
import bodybuilder from 'bodybuilder';
import debounce from 'lodash.debounce';

export const DEFAULT_INDEX = '_all';
export const MAX_SIZE = 10000;
export const SEARCHES_QUEUE = new Set();

export const getExtentWithPadding = (map, { top, left, width, height }) => {
  const topLeft = map.unproject([left, top]).toArray();
  const bottomRight = map.unproject([width + left, height + top]).toArray();
  return [topLeft, bottomRight];
};

export const getExtent = (map, padding) => {
  if (padding) return getExtentWithPadding(map, padding);
  const [[lngMin, latMin], [lngMax, latMax]] = map.getBounds().toArray();
  return [[lngMin, latMax], [lngMax, latMin]];
};

/**
 * Construct an object representing the current properties to search on.
 *
 * @param properties {{}} Filter properties
 * @param form {{}} Schema form containing filters
 * @param key {string} Property key
 * @return {{}} An object usable by searchService (Search.search() or Search.msearch())
 */
export const getSearchParamFromProperty = (properties, form, key) => {
  const { type, values } = form.find(({ property }) => property === key) || {};
  if (properties[key]) {
    // First case, we have a range, the search should be within its bounds
    if (type === 'range') {
      return {
        [key]: {
          type: 'range',
          value: { min: properties[key][0], max: properties[key][1] },
        },
      };
    }
    // Else we have a single or many, the form depends on if we provided values
    if (values) {
      // Then its a keyword, because we already know the values
      return {
        [`${key}.keyword`]: {
          type: 'term',
          value: properties[key],
        },
      };
    }
    // Or else its a "full text" search
    return { [key]: properties[key] };
  }
  return {};
};

/**
 * Build an Elastic Search query
 * @param {String} query A literal query
 * @param {Array} boundingBox Boundingbox to restrict results
 * @param {Number} size Max number of results to fetch
 * @param {Array} properties Array of properties.
 *   Each property should be a key/value association. Value may be a single
 *   value or an object describing the value and type
 * @param {String[]} include List of fields to include in results hits
 * @param {String[]} exclude List of fields to exclude in results hits.
 * @param {Object[]} Array of aggregations definition.
 *   Each aggregation will take a type, a field, a name and options.
 *    See https://bodybuilder.js.org/docs/#aggregation
 */
export const buildQuery = ({
  query = '',
  boundingBox,
  size = MAX_SIZE,
  properties/* = { propName: value }, { propName: { value, type: 'term'} } */,
  include,
  exclude,
  aggregations/* = [{ type, field, name, options }] */,
}) => {
  const body = bodybuilder();
  body.size(size);
  if (query) {
    body.query('query_string', 'query', query.split(/\s+/).map(subquery => `*${subquery}*`).join(' AND '));
  }

  if (properties) {
    Object.keys(properties).forEach(property => {
      const rawValue = properties[property];
      const { value, type = 'match' } = (typeof rawValue === 'object' && !Array.isArray(rawValue))
        ? rawValue
        : { value: rawValue };
      if (value) {
        if (typeof value.min === 'number' && typeof value.max === 'number') {
          const { min, max } = value;
          body.filter('range', property, { gte: +min, lte: +max });
        } else if (Array.isArray(value)) {
          body.filter('bool', q => q
            .orFilter('terms', property, value));
        } else {
          const values = [value];
          values.forEach(val => {
            if (type === 'match' && typeof val === 'string') {
              body.filter('bool', q => q
                .orFilter(type, property, val)
                .orFilter('wildcard', property, `*${val}*`));
            } else {
              body.filter(type, property, val);
            }
          });
        }
      }
    });
  }

  if (boundingBox) {
    body.filter('geo_shape', 'geom', {
      shape: {
        type: 'envelope',
        coordinates: boundingBox,
      },
    });
  }

  if (aggregations) {
    aggregations.forEach(({ type = 'terms', field, options, name }) =>
      body.aggregation(type, field, options, name));
  }

  const bodyBuild = body.build();

  if (include || exclude) {
    const sourceAttr = '_source';
    bodyBuild[sourceAttr] = {};
    if ((include && !include.length)) {
      bodyBuild[sourceAttr] = false;
    } else {
      if (include) {
        bodyBuild[sourceAttr].include = include;
      }
      if (exclude) {
        bodyBuild[sourceAttr].exclude = exclude;
      }
    }
  }

  return bodyBuild;
};

export class Search {
  constructor (host) {
    this.host = host;
  }

  set host (host) {
    this.client = new elasticsearch.Client({ host });
  }

  async search (query = {/*
    query = '',
    boundingBox,
    properties,
    include,
    size = MAX_SIZE,
    index = DEFAULT_INDEX,
  */}) {
    const body = buildQuery(query);
    const action = {
      body,
    };
    const promise = new Promise(resolve => {
      action.resolve = resolve;
    });

    SEARCHES_QUEUE.add(action);
    this.batchSearch();
    return promise;
  }

  async msearch (queries = [/* {
      query: String,q
      include: String,
      properties: {
        property: String,
        //â€¦
      }
    } */]) {
    if (!queries.length) throw new Error('`queries` must not be empty');

    const searches = queries
      .map(({ size = MAX_SIZE / queries.length, ...query }) => buildQuery({ size, ...query }))
      .reduce((body, query) => [...body, { }, query],
        []);
    return this.client.msearch({ body: searches });
  }

  batchSearch = debounce(async () => {
    const [bodies, resolves] = Array
      .from(SEARCHES_QUEUE.values())
      .reduce(([allBodies, allResolves], { body, resolve }) =>
        [[...allBodies, body], [...allResolves, resolve]], [[], []]);
    const batchBody = bodies
      .reduce((body, query) => [...body, { }, query],
        []);
    SEARCHES_QUEUE.clear();
    const { responses } = await this.client.msearch({ body: batchBody });
    resolves.forEach((resolve, index) => resolve(responses[index]));
  }, 500)
}

export default new Search();
