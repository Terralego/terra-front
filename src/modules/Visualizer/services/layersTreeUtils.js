
import searchService, { MAX_SIZE } from './search';

import { PREFIX_SOURCE } from '../../Map/services/cluster';

export const INITIAL_FILTERS = new Map();

export const isCluster = (source, layerId) => !!source.match(new RegExp(`^${layerId}-${PREFIX_SOURCE}-[0-9]+`));

/**
 * Returns a flattened map of layers state from a layers tree
 *
 * @param {object} layersTree The layers tree config object
 * @param {string|string[]} layers Active layer(s) from hash
 * @param {string} table Active table from hash (layer id)
 * @return {Map} A reduced layer tree state
 */
export const initLayersStateAction = (layersTree, { layers: hashLayers, table } = {}) => {
  const layersTreeState = new Map();

  function reduceLayers (group, map) {
    return group.reduce((layersStateMap, layer) => {
      const { initialState = {}, layers: [layerId] = [] } = layer;
      if (layer.group) {
        return reduceLayers(layer.layers, layersStateMap);
      }
      initialState.opacity = initialState.opacity === undefined
        ? 1
        : initialState.opacity;

      if (hashLayers) {
        initialState.active = (
          Array.isArray(hashLayers) && hashLayers.includes(layerId))
          || hashLayers === layerId;
      }

      if (table && table === layerId) {
        initialState.table = true;
      }

      layersStateMap.set(layer, {
        active: false,
        opacity: 1,
        ...initialState,
      });

      return layersStateMap;
    }, map);
  }

  return reduceLayers(layersTree, layersTreeState);
};

export function setGroupLayerStateAction (layer, layerState, prevLayersTreeState) {
  const { layers } = layer;
  let layersTreeState = prevLayersTreeState;
  const defaultIndex = Math.max(0, layers.findIndex(({ default: isDefault }) => isDefault));

  layers.forEach((sublayer, index) => {
    const newLayerState = { ...layerState };
    if (newLayerState.active && index !== defaultIndex) {
      newLayerState.active = false;
    }
    // One must be before the other
    // eslint-disable-next-line no-use-before-define
    layersTreeState = setLayerStateAction(sublayer, newLayerState, layersTreeState);
  });
  return layersTreeState;
}

/**
 * Compute new `newLayersTreeState` from `prevLayersTreeState`
 * with updated `layerState` for `layer`. If we want to `reset` state.
 * Do not update state.
 */
export const setLayerStateAction = (layer, layerState, prevLayersTreeState, reset) => {
  if (layer.group) return setGroupLayerStateAction(layer, layerState, prevLayersTreeState);

  const newLayersTreeState = new Map(prevLayersTreeState);
  // Clone layer current state
  const prevLayerState = newLayersTreeState.get(layer);
  if (!prevLayerState) return prevLayersTreeState;

  const newLayerState = { ...layerState };

  if (newLayerState.table) {
    // Easiest to to read as transform Map in Array and run a .map() on it
    // eslint-disable-next-line no-param-reassign
    newLayersTreeState.forEach(layState => { layState.table = false; });
  }
  newLayersTreeState.set(layer, {
    ...reset ? {} : prevLayerState,
    ...newLayerState,
  });
  return newLayersTreeState;
};

export function layersTreeStatesHaveChanged (layersTreeState, prevLayersTreeState, fields = ['active']) {
  return !fields.reduce((all, field) =>
    all && Array.from(layersTreeState).reduce((suball, [layer, state = {}]) =>
      suball && state[field] === (prevLayersTreeState.get(layer) || {})[field],
    all),
  true);
}

export const filterLayersStatesFromLayersState = (
  layersTreeState = [],
  check = ({ active }) => !!active,
) => Array
  .from(layersTreeState)
  .reduce((allLayers, [layer, state]) =>
    ((check(state)) ? [...allLayers, [layer, state]] : allLayers), []);

export const filterLayersFromLayersState = (layersTreeState, check) =>
  filterLayersStatesFromLayersState(layersTreeState, check)
    .filter(([{ filters: { layer } = {} }]) => layer)
    .map(([{ filters: { layer } }]) => layer);

export const hasTable = layersTreeState =>
  filterLayersFromLayersState(layersTreeState, ({ table }) => table).length > 0;

export const hasWidget = layersTreeState =>
  filterLayersStatesFromLayersState(layersTreeState, ({ widgets = [] }) => widgets.length > 0)
    .length > 0;

const filterFromFeatureIds = ids => {
  if (!Array.isArray(ids)) {
    return null;
  }

  if (!ids.length) {
    return false;
  }

  return ['match', ['get', '_id'], ids, true, false];
};

export const filterFeatures = (
  map,
  features/* = [ { layer: String, features: [id1, id2, â€¦] } ] */,
  layersTreeState,
) => {
  Array.from(layersTreeState).forEach(([{
    layers,
    filters: { layer } = {},
  }, {
    active,
  }]) => {
    if (!active || !layers) return;
    const layerFeatures = features
      .find(({ layer: fLayer }) => fLayer === layer);
    const { features: ids } = layerFeatures || {};

    const filter = filterFromFeatureIds(ids);

    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);
      if (!paintLayer) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }

      if (!INITIAL_FILTERS.has(layerId)) {
        INITIAL_FILTERS.set(layerId, map.getFilter(layerId));
      }
      map.setFilter(layerId, layerFeatures ? filter : INITIAL_FILTERS.get(layerId));
    });
  });
};

export const resetFilters = (map, layersTreeState) => {
  Array.from(layersTreeState).forEach(([{
    layers = [], filters,
  }]) => {
    if (!filters) return;

    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);

      if (!map.getLayer(layerId)) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }
      if (!INITIAL_FILTERS.has(layerId)) return;
      map.setFilter(layerId, INITIAL_FILTERS.get(layerId));
    });
  });
};

const flattenLayersTreeLayersId = layersTree => layersTree.reduce((all, { group, layers }) => [
  ...all,
  ...(group ? flattenLayersTreeLayersId(layers) : layers),
], []);

export const sortCustomLayers = (customLayers, layersTree) => {
  const newCustomLayers = [...customLayers];
  const flattenLayersTree = flattenLayersTreeLayersId(layersTree);
  flattenLayersTree.forEach(layerId => {
    const pos = newCustomLayers.findIndex(({ id }) => id === layerId);
    if (pos > -1) {
      newCustomLayers.push(...newCustomLayers.splice(pos, 1));
    }
  });
  // Reverse order as we want first layer come over next layer
  return newCustomLayers.reverse();
};

/**
 * Fetch property values list
 * @param {String} layer Layer id to fetch
 * @param {Object} property Property to fetch
 * @return Mixed[]
 */
export const fetchPropertyValues = async (layer, { property }) => {
  const results = await searchService.search({
    index: layer,
    aggregations: [{
      type: 'terms',
      field: `${property}.keyword`,
      name: 'values',
      options: {
        size: MAX_SIZE,
      },
    }],
    size: 0,
  });
  const { aggregations: { values: { buckets } } } = results;
  return buckets.map(({ key }) => key);
};

/**
 * Fetch property's min and max values
 * @param {String} layer Layer id to fetch
 * @param {Object} property Property to fetch
 * @return { min: Number, max: Number }
 */
export const fetchPropertyRange = async (layer, { property }) => {
  const results = await searchService.search({
    index: layer,
    aggregations: [{
      type: 'max',
      field: `${property}`,
      name: 'max',
    }, {
      type: 'min',
      field: `${property}`,
      name: 'min',
    }],
    size: 0,
  });
  const { aggregations: { min: { value: min } = {}, max: { value: max } = {} } } = results;

  return { min: min && Math.round(min), max: max && Math.round(max) };
};

/**
 * Transform a layersTree into a flat array of layers
 */
const flattenLayersTreeGroup = tree => {
  const layers = [];

  tree.forEach(treeNode => {
    if (treeNode.label) {
      layers.push(treeNode);
    }

    if (treeNode.layers) {
      layers.push(...flattenLayersTreeGroup(treeNode.layers));
    }
  });

  return layers;
};

/**
 * Story type view needs a specfic format.
 */
export const layersTreeToStory = layersTree => {
  const arrayOfLayers = flattenLayersTreeGroup(layersTree);

  const story = arrayOfLayers.reduce(({
    beforeEach: [beforeEachConfig],
    slides,
  }, {
    label: title,
    content = '',
    layers = [],
    ...layerAttrs
  }) => ({
    beforeEach: [{
      ...beforeEachConfig,
      layers: [
        ...beforeEachConfig.layers,
        ...layers.filter(layer => typeof layer === 'string'),
      ],
    }],
    slides: [
      ...slides, {
        title,
        content,
        layouts: [{
          layers: layers.filter(layer => typeof layer === 'string'),
          active: true,
        }],
        ...layerAttrs,
      }],
  }), {
    beforeEach: [{
      layers: [],
      active: false,
    }],
    slides: [],
  });

  return story;
};

export default {
  initLayersStateAction,
  setLayerStateAction,
  filterLayersFromLayersState,
  hasTable,
  hasWidget,
  filterFeatures,
  resetFilters,
  sortCustomLayers,
  fetchPropertyValues,
  fetchPropertyRange,
  layersTreeToStory,
};
