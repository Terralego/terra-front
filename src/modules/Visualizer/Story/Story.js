import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Button, Intent } from '@blueprintjs/core';
import { toggleLayerVisibility } from '../../Map/services/mapUtils';

import './styles.scss';

const toggleLayers = map => ({ layers, active }) =>
  layers.forEach(layerId =>
    toggleLayerVisibility(map, layerId, active ? 'visible' : 'none'));

const resetLayers = map => ({ layers = [] }) => {
  const { layers: allLayers } = map.getStyle();
  toggleLayers(map)({
    layers: allLayers
      .filter(({ id }) => layers.includes(id))
      .map(({ id }) => id),
    active: false,
  });
};

export class Story extends React.Component {
  static propTypes = {
    story: PropTypes.shape({
      beforeEach: PropTypes.arrayOf(PropTypes.shape({
        layers: PropTypes.arrayOf(PropTypes.string),
        active: PropTypes.bool,
      })),
      slides: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        content: PropTypes.string,
        layouts: PropTypes.arrayOf(PropTypes.shape({
          layers: PropTypes.arrayOf(PropTypes.string),
          active: PropTypes.bool,
        })),
        legends: PropTypes.arrayOf(PropTypes.shape({
          title: PropTypes.string,
          items: PropTypes.arrayOf(PropTypes.shape({
            label: PropTypes.string,
            color: PropTypes.string,
          })),
        })),
      })),
    }).isRequired,
  }

  state = {
    step: 0,
  }

  componentDidMount () {
    this.initMap();
  }

  componentDidUpdate ({ map: prevMap }, { step: prevStep }) {
    const { map } = this.props;
    const { step } = this.state;

    if (map !== prevMap) {
      this.initMap();
    }

    if (map !== prevMap
        || step !== prevStep) {
      this.displayStep();
    }
  }

  prevStep = () => {
    const { story: { slides } } = this.props;
    this.setState(({ step: prevStep }) => ({ step: prevStep === 0
      ? slides.length - 1
      : prevStep - 1 }));
  }

  nextStep = () => {
    const { story: { slides } } = this.props;
    this.setState(({ step: prevStep }) => ({ step: prevStep + 1 === slides.length
      ? 0
      : prevStep + 1 }));
  }

  initMap () {
    const { map } = this.props;

    if (!map) return;

    this.displayStep();
  }

  displayStep () {
    const { map, story: { beforeEach = [], slides }, setLegends } = this.props;
    const { step } = this.state;

    if (!map) return;

    const { layouts = [], legends } = slides[step];
    beforeEach.forEach(resetLayers(map));
    layouts.forEach(toggleLayers(map));
    setLegends(legends);
  }

  render () {
    const { story: { slides } } = this.props;
    const { step } = this.state;
    const { nextStep, prevStep } = this;
    const { title, content } = slides[step];
    const displayPrevButton = step > 0;

    return (
      <div className="storytelling">
        <div className="storytelling__content">
          <h2>{title}</h2>
          <div
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{ __html: content }}
          />
        </div>
        <div className={classnames('storytelling__buttons', { 'storytelling__buttons--justified': displayPrevButton })}>
          {displayPrevButton &&
            <Button icon="chevron-left" onClick={prevStep} large>Précédent</Button>}

          <Button
            rightIcon={step === slides.length - 1 ? undefined : 'chevron-right'}
            icon={step === slides.length - 1 ? 'step-backward' : undefined}
            large
            intent={Intent.PRIMARY}
            onClick={nextStep}
          >
            {step === 0 && 'Démarrer'}
            {step > 0 && step < slides.length - 1 && 'Suivant'}
            {step === slides.length - 1 && 'Recommencer'}
          </Button>
        </div>
        &nbsp;
      </div>
    );
  }
}

export default Story;
