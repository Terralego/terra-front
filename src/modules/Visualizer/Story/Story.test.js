import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { toggleLayerVisibility } from '../../Map/services/mapUtils';

import Story from './Story';

jest.mock('../../Map/services/mapUtils', () => ({
  toggleLayerVisibility: jest.fn(),
}));

const story = {
  beforeEach: [{
    layers: [
      'foo-*',
      'bar',
    ],
    active: false,
  }],
  slides: [{
    title: 'Slide 1',
    content: `
  <p>
    Content 1
  </p>
`,
    layouts: [{
      layers: ['foo-1a', 'foo-1b'],
      active: true,
    }],
  }, {
    title: 'Slide 2',
    content: `
<p>
  Content 2
</p>
`,
    layouts: [{
      layers: ['foo-1a', 'foo-1b'],
      active: true,
    }],
  }, {
    title: 'Slide 3',
    content: `
<p>
  Content 3
</p>
`,
    layouts: [{
      layers: ['foo-2'],
      active: true,
    }],
    legends: [{
      title: 'Legend 1',
      items: [
        { label: 'a', color: '#ff8c00' },
        { label: 'b', color: '#965096' },
        { label: 'c', color: '#ff78a0' },
        { label: 'd', color: '#003c82' },
        { label: 'e', color: '#f5000a' },
        { label: 'f', color: '#c8be00' },
        { label: 'g', color: '#90ff77' },
        { label: 'h', color: '#007dff' },
        { label: 'i', color: '#1e6414' },
      ],
    }],
  }, {
    title: 'Slide 4',
    content: `
<p>
  Content 4
</p>
    `,
    layouts: [{
      layers: ['foo-3'],
      active: true,
    }],
    legends: [{
      title: 'Legend 2',
      items: [
        { label: '1', color: '#b91a1c' },
        { label: '2', color: '#b3de69' },
        { label: '3', color: '#1f78b4' },
        { label: '4', color: '#a6a5ae' },
        { label: '5', color: '#ec7c30' },
        { label: '6', color: '#fedb2a' },
        { label: '7', color: '#8fa9db' },
        { label: '8', color: '#9e480d' },
      ],
    }],
  }, {
    title: 'Slide 5',
    content: `
<p>
  Content 5
</p>
`,
    layouts: [{
      layers: ['foo-4'],
      active: true,
    }],
    legends: [{
      title: 'Legend 3',
      items: [
        { label: 'a', color: '#fcbba1', shape: 'circle', radius: 5 },
        { label: 'aa', color: '#fc9272', shape: 'circle', radius: 10 },
        { label: 'aaa', color: '#fb6a4a', shape: 'circle', radius: 15 },
        { label: 'aaaa', color: '#ef3b2c', shape: 'circle', radius: 20 },
        { label: 'aaaaa', color: '#cb181d', shape: 'circle', radius: 25 },
        { label: 'aaaaaa', color: '#99000d', shape: 'circle', radius: 30 },
      ],
    }],
  }, {
    title: 'Slide 6',
    content: `
<p>
  Content 6
</p>
`,
    layouts: [{
      layers: ['bar'],
      active: true,
    }],
  }],
};

it('should render', () => {
  const { asFragment } = render(
    <Story story={story} />,
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should call callback on mount', () => {
  const setLegends = jest.fn();
  render(
    <Story
      setLegends={setLegends}
      story={story}
      map={{
        getStyle: jest.fn(() => ({ layers: [{ id: 'foo-1a' }] })),
        setLayoutProperty: jest.fn(),
      }}
    />,
  );
  expect(toggleLayerVisibility).toHaveBeenCalled();
  expect(setLegends).toHaveBeenCalled();
});

it('should set nextStep', () => {
  const { container, getByText } = render(
    <Story story={story} />,
  );
  const nextButton = container.querySelector('.bp3-intent-primary');
  fireEvent.click(nextButton);
  const title = getByText('Slide 2');
  expect(title).toBeTruthy();
});

it('should set nextStep at the last slide and loop to the first step', () => {
  const { container, getByText } = render(
    <Story story={story} />,
  );
  const nextButton = container.querySelector('.bp3-intent-primary');
  story.slides.forEach(() => {
    fireEvent.click(nextButton);
  });
  const title = getByText('Slide 1');
  expect(title).toBeTruthy();
});

it('should not have prevStep button in the first step', () => {
  const { container } = render(
    <Story story={story} />,
  );
  const nextButton = container.querySelector('.bp3-intent-primary');
  story.slides.forEach((_, index) => {
    const buttons = container.querySelectorAll('.bp3-button');
    expect(buttons.length).toBe((index === 0) ? 1 : 2);
    fireEvent.click(nextButton);
  });
});

it('should set prevStep', () => {
  const { container } = render(
    <Story story={story} />,
  );
  const nextButton = container.querySelector('.bp3-intent-primary');
  const title = container.querySelector('h2');
  fireEvent.click(nextButton);
  expect(title.textContent).toBe('Slide 2');
  const prevButton = container.querySelector('.bp3-button');
  fireEvent.click(prevButton);
  expect(title.textContent).toBe('Slide 1');
});

it('should not display buttons if there is only one step', () => {
  const { container } = render(
    <Story story={{ ...story, slides: [story.slides[0]] }} />,
  );
  const buttonContainer = container.querySelector('.storytelling__buttons');
  expect(buttonContainer).toBe(null);
});
