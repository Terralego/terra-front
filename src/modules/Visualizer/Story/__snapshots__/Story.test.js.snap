// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="storytelling"
  >
    <div
      class="storytelling__content"
    >
      <h2>
        Slide 1
      </h2>
      <div>
        
  
        <p>
          
    Content 1
  
        </p>
        

      </div>
    </div>
    <div
      class="storytelling__buttons"
    >
      <button
        class="bp3-button bp3-large bp3-intent-primary"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Start
        </span>
        <span
          class="bp3-icon bp3-icon-chevron-right"
          icon="chevron-right"
        >
          <svg
            data-icon="chevron-right"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              chevron-right
            </desc>
            <path
              d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
